doc/rtd/howto/return_codes.rst

```diff
@@ -0,0 +1,172 @@
+  previously reported a status of "done".
+
+.. _return_codes_improvements
+
+Efforts to improve cloud-init
+-----------------------------
+
+Several changes have been introduced to cloud-init to address the pain
+points described above.
+
+Jsonschema
+^^^^^^^^^^
+
+Cloud-init has defined a jsonschema which fully documents the user-data
+cloud-config. This jsonschema may be used in several different ways:
+
+Text editor integration
+"""""""""""""""""""""""
+
+Thanks to `yamlls<yaml-language-server>`, cloud-init's jsonschema may be
+used for yaml syntax checking, warnings when invalid keys are used, and
+even autocompletion. Several different text editors are capable of this.
+See this `blog post on configuring this for neovim`_, or 
+
+# TODO someone who knows about vscode add some details
+
+Cloud-init schema subcommand
+""""""""""""""""""""""""""""
+
+The cloud-init package includes a cloud-init subcommand, `cloud-init schema`
+which uses the schema to validate either the configuration passed to the
+instance that you are running the command on, or to validate an arbitrary
+text file containing a configuration.
+
+Nuanced Status Reporting
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+In release 23.4, cloud-init introduced extended status reporting to users.
+This makes it easier for users to identify problems that didn't result in a
+crash.
+
+Extended status and recoverable errors
+""""""""""""""""""""""""""""""""""""""
+
+If a user's configuration was invalid, or if the operating system or cloud
+environment experienced some error that prevented cloud-init from configuring
+the instance, or if cloud-init internally experienced an error, this status
+is now visible via :code:`cloud-init status --long` after the `extended_status`
+key. Additionally, the cause of this error will be visible after the
+`recoverable_errors` key.
+
+Return codes
+""""""""""""
+
+Cloud-init historically used two return codes from the `cloud-init status`
+subcommand: 0 to indicate success and 1 to indicate failure. These return codes
+lacked nuance. Return code 0 (success) included the in-between when something
+went wrong, but cloud-init was able to finish.
+
+Many users of cloud-init run :code:`cloud-init status --wait` and expect that
+when complete, cloud-init has finished. Since cloud-init is not guaranteed to
+succeed, users should also be check the return code of this command.
+
+A backwards-incompatible change was recently introduced: errors that did not
+crash cloud-init now have an exit code of 2. After this change, exit 1 still
+means that cloud-init crashed, but exit code 0 now more correctly means that
+cloud-init succeeded. Anyone that previously checked for exit code 0 should
+probably update their assumptions in one of the following two ways:
+
+Users that wish to take advantage of cloud-init's new error reporting
+capabilities should check for exit code of 2 from :code:`cloud-init status`.
+An example of this:
+
+.. code-block:: python
```

Comment:
Shell might be a better example here just as that's the general the lowest common denominator for scripting.

Reply:
I was originally going to use shell, but it doesn't lend well to making use of the machine-readable output available in cloud-init, which real languages would want to make use of, which is why I went with Python.